File: Google Kick Start 2020/A - Plates/main.cpp
#include <iostream>
#include <string>
#include <string.h>
#include <vector>
#include <algorithm>
//#include <bits/stdc++.h>

using namespace std;

int N, K, P;
int cache[50][50*30];

int rec(int i, int taken, auto &plates) {
    if (i >= N || taken >= P)
        return 0;
    if (cache[i][taken] != -1) return cache[i][taken];
    int ans = rec(i+1, taken, plates);
    int cur = 0;
    for (int j = 0; j < K && taken+j+1 <= P; j++) {
        cur += plates[i][j];
        ans = max(ans, cur + rec(i+1, taken+j+1, plates));
    }
    cache[i][taken] = ans;
    return ans;
}

int main() {
    int numtc;
    cin >> numtc;
    for (int tc = 1; tc <= numtc; tc++) {
        cin >> N >> K >> P;
        vector<vector<int>> plates(N, vector<int> (K, 0));
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < K; j++) {
                int p;
                cin >> p;
                plates[i][j] = p;
            }
        }
        memset(cache, -1, 50*50*30*4);
        cout << "Case #" << tc << ": " << rec(0, 0, plates) << endl;
    }
}

File: Google Kick Start 2020/A - Plates/main.js

function rec(stackIndex, numTaken) {
    if (stackIndex >= stacks.length)
        return 0
    if (cache[stackIndex][numTaken] != undefined)
        return cache[stackIndex][numTaken]
    let ans = rec(stackIndex+1, numTaken)
    let cur = 0
    for (let i = 0; i < stacks[stackIndex].length && numTaken+i+1 <= numToTake; i++) {
        cur += stacks[stackIndex][i]
        ans = Math.max(ans, cur + rec(stackIndex+1, numTaken+i+1))
    }
    cache[stackIndex][numTaken] = ans
    return ans
}

lines = require('fs').readFileSync(0, 'utf8').split('\n')
numTestCases = parseInt(lines.shift())
for (let i = 0; i < numTestCases; i++) {
    let inputs = lines.shift().split(' ').map(x=>parseInt(x))
    var stacks = Array(inputs[0]).fill().map(() => lines.shift().split(' ').map(x=>parseInt(x)))
    var numToTake = inputs[2]
    var cache = Array(stacks.length).fill().map(() => [])
    let ans = rec(0, 0)
    console.log(`Case #${i+1}: ${ans}`)
}
